smarts/waymo/waymo_data_extraction.py

```diff
@@ -0,0 +1,266 @@
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import argparse
+import logging
+import os
+import pickle
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import List, Optional
+
+from PIL import Image
+from smarts.core.agent_interface import AgentInterface
+from smarts.core.controllers import ActionSpaceType, ControllerOutOfLaneException
+from smarts.core.scenario import Scenario
+from smarts.core.smarts import SMARTS
+from smarts.env.hiway_env import HiWayEnv
+from smarts.env.wrappers.format_obs import FormatObs
+from smarts.sstudio import build_scenario
+from smarts.zoo.agent_spec import AgentSpec
+
+logger = logging.getLogger(__file__)
+logger.setLevel(logging.INFO)
+
+AGENT_SPEC = AgentSpec(
+    interface=AgentInterface(
+        accelerometer=True,
+        action=ActionSpaceType.Continuous,
+        neighborhood_vehicles=False,
+        rgb=True,
+        road_waypoints=False,
+        waypoints=False,
+    ),
+)
+
+
+def _scenario_code(dataset_path, scenario_id):
+    return f"""from pathlib import Path
+
+from smarts.sstudio import gen_scenario
+from smarts.sstudio import types as t
+
+dataset_path = "{dataset_path}"
+scenario_id = "{scenario_id}"
+
+traffic_histories = [
+    t.TrafficHistoryDataset(
+        name=f"waymo",
+        source_type="Waymo",
+        input_path=dataset_path,
+        scenario_id=scenario_id,
+    )
+]
+
+gen_scenario(
+    t.Scenario(
+        map_spec=t.MapSpec(
+            source=f"{{dataset_path}}#{{scenario_id}}", lanepoint_spacing=1.0
+        ),
+        traffic_histories=traffic_histories,
+    ),
+    output_dir=Path(__file__).parent,
+)
+"""
+
+
+def _record_data(
+    smarts,
+    collected_data,
+    output_dir,
+    current_vehicles,
+    off_road_vehicles,
+    selected_vehicles,
+    std_obs_wrapper,
+):
+    # Attach sensors to each vehicle
+    valid_vehicles = (current_vehicles - off_road_vehicles) & selected_vehicles
+    for veh_id in valid_vehicles:
+        try:
+            smarts.attach_sensors_to_vehicles(AGENT_SPEC.interface, {veh_id})
+        except ControllerOutOfLaneException:
+            logger.warning(f"{veh_id} out of lane, skipped attaching sensors")
+            off_road_vehicles.add(veh_id)
+
+    # Get observations from each vehicle and record them
+    obs, _, _, _ = smarts.observe_from(list(valid_vehicles))
+    obs = std_obs_wrapper.observation(obs)
```

Comment:
This is the only part that still feels pretty hacky. I didn't think it was worth refactoring the `FormatObs` class for this PR, but we may want to in the future. Let me know if there is a better way to achieve this for this PR.

Reply:
+ Why did we not just use the `smarts.sensors.Observation` directly?
+ Why did we use the `StdObs` from `FormatObs` wrapper here to reformat the observation?
+ By using `FormatObs` wrapper here, we would be making this standard format mandatory for all users.

Reply:
Do we have mission goals here?